### ISSUE TITLE
Rule: Unnecessary `let`

### ISSUE THREAD

**Issue by Nycto** [arturbosch/detekt#520]
I've seen this one come through in a few code reviews, I would love to see it linted away:

```
nullableVal?.let { it.method() }
```

They should just be calling `nullableVal?.method()` directly on the parent object

**Comment by schalkms** [comment]
Good point!

**Comment by 3flex** [comment]
I've seen a couple other variations:
`this.let { return it }` should probably be `return this`
`path.let{ File(it) }` should probably be `File(path)`

**Comment by arturbosch** [comment]
@3flex you mean if let is used on non-nullable-types?

**Comment by 3flex** [comment]
More that it should flag usage of `let` where there's a single expression in the block that references `it`, assuming that using it that way is always unnecessary?

**Comment by Mauin** [comment]
If we have a rule that flags unnecessary usages of `let` we should cover both the nullable and the non-nullable case. Also instead of `it` the lambda could have a named parameter that we would need to consider.

`something.let { myparameter -> myparameter.doSomething() }`

**Comment by Mishkun** [comment]
> `path.let{ File(it) }` should probably be `File(path)`

this variation is perfectly Ok, it the purpose of let in the first place: to pipe the operations and transform lisp-like code 
```
foo(bar(baz(barre(bizarre(x)))))
```
to more Java-ish chained calls:
```
x.let(::bizarre).let(::barre).let(::baz).let(::bar).let(::foo)
```
